bbcondeparser/html_tags.py

```diff
@@ -358,108 +343,235 @@ class BaseHTMLRenderTreeParser(BaseTreeParser):
+            self.close_paragraph()
 
-    if convert_paragraphs and inside_paragraph:
-        rendered_children.append('</p>')
-        inside_paragraph = False
+        self.handle_tree()
+        self.append_tree(self._node)
 
-    return ''.join(rendered_children)
+    def handle_renderable_text(self):
+        if self.is_converting_paragraphs:
+            if not self.is_inside_paragraph:
+                self.open_paragraph()
+            else:
+                self.append_paragraph(self._node)
+
+        else:
+            self.append_tree(self._node)
+
+    def handle_newline(self):
+        # Open/close a paragraph
+        if self._node.count >= 2 and self.is_converting_paragraphs:
+
+            # Close paragraph, wrap in tag, add as node
+            if self.is_inside_paragraph and self._paragraph_tree:
+                self.close_paragraph()
+
+            # Open paragraph and scope. Dont need to add node.
+            else:
+                self.open_paragraph(include_node=False)
+
+        else:
+            if self.is_converting_newlines:
+                # we dont want to put <br/> in between blocks so
+                # we actually only want to put them in paragraphs
+                # if we're converting or anywhere if we're not
+                # converting paragraphs
+                if self.is_inside_paragraph:
+                    self.append_paragraph(self._node)
+                elif not self.is_converting_paragraphs:
+                    self.append_tree(self._node)
+                else:
+                    # Not removing and not converting so change
+                    # back to NewlineText
+                    new_node = NewlineText(self._node.text)
+                    new_node.count = self._node.count
+                    self.append_tree(new_node)
+
+            elif not self.is_removing_newlines:
+                # Not removing and not converting so change back
+                # to NewlineText
+                new_node = NewlineText(self._node.text)
+                new_node.count = self._node.count
+                if self.is_inside_paragraph:
+                    self.append_paragraph(new_node)
+                else:
+                    self.append_tree(new_node)
+
+    def handle_tree(self):
+
+        self.stack_push()
+
+        for node in self._node.tree:
+            self._node = node
+
+            if is_inline_tag(self._node):
+                self.handle_inline_tag()
+
+            elif is_block_tag(self._node):
+                self.handle_block_tag()
+
+            elif isinstance(self._node, RENDERABLE_TEXT):
+                self.handle_renderable_text()
+
+            elif isinstance(self._node, NewlineText):
+                self.handle_newline()
+
+        if self.is_inside_paragraph and self._paragraph_tree:
+            self.close_paragraph()
+
+        _tree = self._tree
+        self.stack_pop()
+        self._node.tree = _tree
```

Comment:
Need to set `parent` on the children in the tree.

Reply:
ive done this but GH isnt picking it up 8a04b52b6fb7896a15f81cc6a35b2d473fb40f9b